### PR TITLE
Fix AssertionError in MustCallInference::inferOwningField with @MustCallUnknown fields, fixes #6480

### LINKED ISSUE #6480: AssertionError in MustCallInference::inferOwningField with @MustCallUnknown Fields
## Description:
The Checker Framework throws an AssertionError when attempting to infer the `@Owning` annotation for fields with a `@MustCallUnknown` annotation. This occurs during the processing of generic class fields when the MustCallInference method incorrectly handles fields marked as `@MustCallUnknown`, leading to a failure in the assertion due to both @CalledMethods({}) and mustCallValues being empty.

## Steps to Reproduce:

Create a Java file with the following content:
```java
class Generic<T> {
  public T data;

  public Generic(T data) {
    this.data = data;
  }
}

public class GenericClassFieldCrash {
  private void onPacket(Generic foo) {
    String.format("socket received: data '%s'", foo.data);
  }
}
```
Run the Checker Framework (version 3.43.0-SNAPSHOT) on this file with the Resource Leak Checker enabled. Or run the resource leak inference test suite. 
```
./gradlew AinferResourceLeakAjavaTest
```

## Expected Behavior:
The Checker Framework should either correctly infer the `@Owning` annotation for fields marked as `@MustCallUnknown` or conservatively avoid inferring an `@Owning` annotation to prevent an AssertionError.

## Actual Behavior:
The Checker Framework crashes with an AssertionError, indicating a failure to handle `@MustCallUnknown` fields correctly during the inference process.

Stack Trace:
```java
error: SourceChecker.typeProcess: unexpected Throwable (AssertionError) while processing /Users/sanjay/checker-framework/checker/tests/ainfer-resourceleak/non-annotated/GenericClassFieldCrash.java; message: Must-call obligation of owning field data is empty.
  ; The Checker Framework crashed.  Please report the crash.  Version: Checker Framework 3.43.0-SNAPSHOT, branch master, 2024-03-04, commit 7743e5d, dirty=true. 
  Compilation unit: /Users/sanjay/checker-framework/checker/tests/ainfer-resourceleak/non-annotated/GenericClassFieldCrash.java
  Last visited tree at line 10 column 1:
  public class GenericClassFieldCrash {
  Exception: java.lang.AssertionError: Must-call obligation of owning field data is empty.; java.lang.AssertionError: Must-call obligation of owning field data is empty.
  	at org.checkerframework.checker.resourceleak.MustCallInference.inferOwningField(MustCallInference.java:402)
  	at org.checkerframework.checker.resourceleak.MustCallInference.inferOwningForParamOrField(MustCallInference.java:716)
  	at org.checkerframework.checker.resourceleak.MustCallInference.inferOwningForRecieverOrFormalParamPassedToCall(MustCallInference.java:693)
  	at org.checkerframework.checker.resourceleak.MustCallInference.inferOwningFromInvocation(MustCallInference.java:796)
  	at org.checkerframework.checker.resourceleak.MustCallInference.runInference(MustCallInference.java:255)
  	at org.checkerframework.checker.resourceleak.MustCallInference.runMustCallInference(MustCallInference.java:213)
  	......
	  at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	  at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```

## Analysis:
The method `MustCallInference::inferOwningField` incorrectly assumes that a field with a `@MustCallUnknown` annotation can have a non-empty must-call obligation. This leads to an assertion failure when both the called methods annotation (`cmAnno`) and the must-call values (`mustCallValues`) are empty, despite passing the must-call consistency check.

## Proposed Solution:
Modify the `ResourceLeakAnnotatedTypeFactory::isFieldWithNonemptyMustCallValue` method to return `false` for fields whose MC type is `@MustCallUnknown`. This change will conservatively avoid inferring an `@Owning` annotation. Thus, avoiding incorrectly inferring an `@Owning` annotation for fields marked as `@MustCallUnknown`, addressing the assertion error.

## Environment:
Checker Framework Version: 3.43.0-SNAPSHOT
Java Version: JDK 8

### PR DESCRIPTION
**Pull request by iamsanjaymalakar**
Fixes #6480

### CODE REVIEW

File: checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakAnnotatedTypeFactory.java [hunk 1]
```diff
@@ -107,13 +107,26 @@ public ResourceLeakAnnotatedTypeFactory(BaseTypeChecker checker) {
 
   /**
    * Is the given element a candidate to be an owning field? A candidate owning field must have a
-   * non-empty must-call obligation.
+   * non-empty must-call obligation. This method retrieves the {@code @MustCall} annotation of a
+   * field, if any. If the {@code @MustCall} annotation is missing, which implies
+   * {@code @MustCallUnknown}, this method conservatively treats the field as having an empty
+   * must-call obligation and returns false. Otherwise, it checks if the {@code @MustCall} values
```
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakAnnotatedTypeFactory.java:113]: @kelloggm can confirm, but I don't think this is correct.  `@MustCallUnknown` means we don't know what must be called, and a conservative handling would be something like assuming every method must be called, not assuming that the must-call set is empty.

I'm wondering if we could do a narrower fix for the issue just inside inference?  Like an explicit check for a `@MustCallUnknown` type wherever the assertion was failing?

Comment (kelloggm) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakAnnotatedTypeFactory.java:113]: The problem here is that this method is in `ResourceLeakAnnotatedTypeFactory`, but it's intended for inference. It's documentation already makes that clear (e.g., "Is the given element a candidate to be an owning field?"). Given that this method only makes one call to another RLATF instance method, I'd prefer to "fix" this by moving the method to `MustCallInferenceLogic`, making it static, and then passing it the RLATF as a parameter.

Comment (kelloggm) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakAnnotatedTypeFactory.java:113]: (Within the context of inference, "don't infer anything" is the conservative thing to do, and that is how we should handle `@MustCallUnknown`.)

File: checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakAnnotatedTypeFactory.java [hunk 2]
```diff
@@ -107,13 +107,26 @@ public ResourceLeakAnnotatedTypeFactory(BaseTypeChecker checker) {
 
   /**
    * Is the given element a candidate to be an owning field? A candidate owning field must have a
-   * non-empty must-call obligation.
+   * non-empty must-call obligation. This method retrieves the {@code @MustCall} annotation of a
```
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakAnnotatedTypeFactory.java:110]: @iamsanjaymalakar why do we still need changes in this file?

**Review by msridhar — COMMENTED** [review]
Quick initial question

File: checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakAnnotatedTypeFactory.java (hunk 2 above)
Comment (iamsanjaymalakar) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakAnnotatedTypeFactory.java]: @msridhar Yes, I forgot to remove the previous function that is not usud now anymore.

But there are still some change in this file. I've refactored some common codes to a new method, that was used twice in RLATF and my newly added method in MustCallInference. So, I thought my as well refactor some code. If this doesn't work, let me know. I will remove the refactoring.

Comment (mernst) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakAnnotatedTypeFactory.java]: Refactorings are good (thanks!), but should always be in a separate pull request that consists only of the refactoring.  Could you please create such a pull request?

Comment (iamsanjaymalakar) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakAnnotatedTypeFactory.java]: Sure, I can do that.

Comment (iamsanjaymalakar) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakAnnotatedTypeFactory.java]: > Refactorings are good (thanks!), but should always be in a separate pull request that consists only of the refactoring. Could you please create such a pull request?

I've created a seprate PR with only the refactoring here: #6489. @mernst 

Do I need to update this PR after refactoring PR gets merged into the master branch?

Comment (mernst) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakAnnotatedTypeFactory.java]: After the refactoring PR is merged, then merge master into this branch.  There will be a button to do it.  That will reduce the diffs in this PR.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java [hunk 3]
```diff
@@ -380,6 +380,27 @@ private void addOwningToParam(int index) {
     wpi.addDeclarationAnnotationToFormalParameter(methodElt, index, OWNING);
   }
 
+  /**
+   * This method checks if a field is an owning candidate. A field is an owning candidate if it has
+   * a non-empty must-call obligation. Unless, it is a @MustCallUnknown, in which we don't want to
+   * infer anything. So, we conservatively treat it as a non-owning candidate.
```
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:386]: Wording:
```suggestion
   * This method checks if a field is an owning candidate. A field is an owning candidate if it has
   * a non-empty must-call obligation, unless it is {@code @MustCallUnknown}.  For a {@code @MustCallUnknown} field, we don't want to infer anything, so, we conservatively treat it as a non-owning candidate.
```

**Review by msridhar — APPROVED** [review]
Looks great, just one suggestion for comment wording

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java (hunk 3 above)
Comment (iamsanjaymalakar) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java]: @msridhar I've updated the wording.

**Review by msridhar — APPROVED** [review]
LGTM

**Pull request merged**